### ISSUE TITLE
Excessive AWS API calls for simple tasks

### ISSUE THREAD

**Issue by jacksontj** [aws/amazon-vpc-cni-k8s#730]
Lets say I have a node with a `WARM_IP_TARGET` of 5 and a single daemonset running. When I go to deploy that daemonset we'll need to give an IP back to the CNI and then (shortly afterwards) allocate a new one. After just doing this here is what I see as happening:

## scenario before
1. start: `inuse=1 spare=5`
2. ds pod is stopped: `inuse=0 spare=6`
3 "over" ip is freed: `inuse=0 spare=5` **(api call)**
4. ds pod is started, [CNI allocates *all* available IPs to fill ENI](https://github.com/aws/amazon-vpc-cni-k8s/blob/master/ipamd/ipamd.go#L673): `inuse=1 spare=10` **(api call)**
5. CNI frees IPs to get back to `WARM_IP_TARGET`: `inuse=1 spare=5` **(api call)**

Total API calls: 3

So in this very simple scenario (deploying a single pod on a node) we have made 3 API calls and over-allocated by 5 IPs.

It seems that the ipamd IP pool could use some better logic around how it allocates/frees IPs.

I'll propose a few improvements:
- Only allocate as many IPs as necessary to hit IP targets: there' no need to fill an ENI just to turn around and release them
- Only free IPs after they have been idle for some configurable time: in the case that pods are rolling its very likely the IP would be needed anyway, especially in the case where we have other IPs free and available
- only allocate *more* IPs after some configurable time: its also common for deployments to have a surge of pods, if we still have spare IPs available we might want to wait a while as the old pods might release the IPs back.

With these proposed improvements the above example could look like this:

## scenario after
1. start: `inuse=1 spare=5`
2. ds pod is stopped: `inuse=0 spare=6`
3. ds pod is started: `inuse=1 spare=5`

Total API calls: 0

**Comment by mogren** [comment]
Hi @jacksontj! You are correct that ipamd will try to attach all secondary IPs on the ENI at first, then [after 30 seconds deallocate](https://github.com/aws/amazon-vpc-cni-k8s/blob/1726afd4395c6c02f5caab09daf56b447719bb31/ipamd/ipamd.go#L51) the ones that are not needed. The reason for this is an optimization for the quite common use case that new nodes suddenly get a couple of pods scheduled directly, from daemonsets or similar, but then stabilize. 

This can definitely be improved in many ways. For example, having a "tolerance", something like IP target 5, tolerance 2, would allow having a range of 3-7 IPs free before doing any allocating or deallocating. Maybe having a 30% tolerance, with a max of 5 IPs could be a reasonable default?

Also, like you said, we could probably increase the default cooldown time before we free unused IPs.

**Comment by jacksontj** [comment]
> The reason for this is an optimization for the quite common use case that new nodes suddenly get a couple of pods scheduled directly, from daemonsets or similar, but then stabilize.

Shouldn't this use-case be covered by `MINIMUM_IP_TARGET`? That seems like a better way of handling "node just started and needs N IPs".

Your "tolerance" sounds like what I was describing. My thought was basically to have a tolerance + timeout so (in your example) if I have 3 free IPs I won't immediately get more until after some cooldown (or if we drop below that number).

I'd actually argue that we probably don't want to *fill* the ENI (as that use-case should be covered by `MINIMUM_IP_TARGET`) and at most we could fill to the max of our tolerance (7 in this case) to avoid over-allocating and re-releasing the IPs.


> Also, like you said, we could probably increase the default cooldown time before we free unused IPs.

I think we probably need that configurable/increase as well as the addition of a cooldown time for scaling up (as long as we are within the tolerance).


So to summarize our proposal as of now:
- add a "tolerance" window; so we can say something like `WARM_IP_TARGET`=5 +/ 30%
- add a configurable cooldown for adding/removing IPs that would be honored assuming free ips are in the "tolerance" window
- stop "filling" ENIs but instead fill to either `WARM_IP_TARGET` or the max tolerance (to avoid over filling).

**Comment by mogren** [comment]
@jacksontj Sounds good to me. The recent addition of `MINIMUM_IP_TARGET` is definitely a better way to solve the cold start problem.

**Comment by mogren** [comment]
With the added `MINIMUM_IP_TARGET ` and the recent fix in #809 I think we are getting closer to solving this issue. Added #817 for having a configurable reconcile time.

**Comment by mogren** [comment]
`MINIMUM_IP_TARGET` together with the cool down solves this well enough.
